source/bigint.c

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/bigint.h>
+
+#define BASE_BITS 32
+#define NIBBLES_PER_DIGIT ((BASE_BITS) / 4)
+#define LOWER_32_BIT_MASK 0xFFFFFFFF
+#define INT64_MIN_AS_HEX 0x8000000000000000
+
+/*
+ * Working set of invariants:
+ *
+ * (1) Negative zero is illegal
+ * (2) The only time leading (trailing) 0-value digits are allowed is a single instance to represent actual zero.
+ *
+ * Other proposals:
+ *
+ * (1) Functionally immutable API (no visible side-affects)
+ * (2) params may be temporarily altered during computation before being restored in order to prevent pointless copying
+ *   (most often around sign negation and operator switching)
+ *
+ * Dev ideas:
+ *
+ * (1) Internal arithmetic ops may use a "raw" interface to allow for subsequence operations used in multiply/divide.
+ */
+
+void aws_bigint_clean_up(struct aws_bigint *bigint) {
+    aws_array_list_clean_up(&bigint->digits);
+    AWS_ZERO_STRUCT(bigint->digits);
+}
+
+static void s_advance_cursor_past_hex_prefix(struct aws_byte_cursor *hex_cursor) {
+    if (hex_cursor->len >= 2) {
```

Comment:
if `hex_cursor == NULL`, then NULL pointer deference, which could be avoided here by adding at the beginning `AWS_PRECONDITION(aws_byte_cursor_is_valid(hex_cursor))`.

Reply:
Should the check be here or is it sufficient to put a precondition on the cursor at the entry point of the public function that uses this? (how much redundancy?)

Reply:
If you added to the entry point, there is no need to add it here as well; however, you must make sure that all entry points that call `s_advance_cursor_past_hex_prefix` contains the `PRECONDITION`. AFAIK, the asserts are only triggered in debug mode (all preconditions are treated as assertions), would it be so bad to add it in both places?

Reply:
Should be fine, this isn't really performance sensitive.